### PR TITLE
Parallelized builds & enabled compile cache

### PR DESCRIPTION
**Pull request by ajayborra**
**Related issues**
https://github.com/salesforce/TransmogrifAI/issues/104

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -6,23 +6,58 @@ before_install:
   - unset _JAVA_OPTIONS
 env:
   global:
-    - GRADLE_OPTS="-Xmx256m"
+    - GRADLE_OPTS='-Xmx256m'
+    - BUILD_CACHE_DIR="$HOME/build-cache"
+    - TESTS_FILE_NAME="$TRAVIS_BUILD_DIR/tests.txt"
+    - TESTS_FILES_PREFIX='tests'
+    - OP_VERSION_PROPS_DISABLED=false
```
Comment (tovbinm) [inline .travis.yml]: `OP_VERSION_PROPS_DISABLED=true`

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,69 +1,58 @@
...
-  build_core:
+          name: Scala style check
+          command: ./gradlew scalaStyle --parallel
+  clone_repo_and_compile_project:
     machine: true
     working_directory: ~/repo
     environment:
-      GRADLE_OPTS: -Xmx256m -Dorg.gradle.daemon=true -Dorg.gradle.caching=true
-      TERM: dumb
+      OP_VERSION_PROPS_DISABLED: false
```
Comment (tovbinm) [inline .circleci/config.yml]: `OP_VERSION_PROPS_DISABLED: true`

File: build.gradle [hunk 3]
```diff
@@ -159,6 +159,16 @@ configure(allProjs) {
             "-language:implicitConversions", "-language:existentials", "-language:postfixOps"
         ]
     }
+    compileScoverageScala {
```
Comment (tovbinm) [inline build.gradle]: Try this instead:
```
tasks.withType(ScalaCompile) {
        configure(scalaCompileOptions.forkOptions) {
            memoryMaximumSize = '1g'
            jvmArgs = ['-XX:MaxMetaspaceSize=256m']
        }
        scalaCompileOptions.additionalParameters = [
           "-Yrangepos", "-feature",
           "-language:implicitConversions", "-language:existentials", "-language:postfixOps"
        ]
    }
    compileScala.scalaCompileOptions.additionalParameters += "-optimize"
```

File: .travis.yml (hunk 1 above)
Comment (ajayborra) [reply .travis.yml]: Was wondering how the local testing passed when false is set. Perhaps [this check](https://github.com/salesforce/TransmogrifAI/blob/master/gradle/version-properties.gradle#L19) is validating if the property is defined (or) not.

File: .travis.yml [hunk 4]
```diff
@@ -6,23 +6,58 @@ before_install:
...
+        ./gradlew --no-daemon reportScoverage
+      name: 'Build Part 3'
+    - script: |
+        export TEST_FILES=`cat "$BUILD_CACHE_DIR/$TESTS_FILES_PREFIX"03`
+        ./gradlew --no-daemon reportScoverage
+      name: 'Build Part 4'
+    - script: |
+        export TEST_FILES=`cat "$BUILD_CACHE_DIR/$TESTS_FILES_PREFIX"04`
+        ./gradlew --no-daemon reportScoverage
+      name: 'Build Part 5'
```
Comment (tovbinm) [inline .travis.yml]: Part 5?

File: build.gradle [hunk 5]
```diff
@@ -173,7 +166,7 @@ configure(allProjs) {
         source = "src/main/scala"
         testSource = "src/test/scala"
     }
-    compileScala.dependsOn(scalaStyle)
+    // compileScala.dependsOn(scalaStyle)
```
Comment (tovbinm) [inline build.gradle]: We only need to disable scalaStyle for CI, but it has to be enabled on local builds.

Comment (ajayborra) [reply build.gradle]: Shall we introduce a variable to distinguish between local vs cloud build env? Alternatively, we can update the contributors guide.

Comment (tovbinm) [reply build.gradle]: TravisCI and CircleCI already expose a special variable `CI=true`

**Pull request merged**